### PR TITLE
Observing lists fix

### LINKED ISSUE #976: Bookmarks becomes Observing List
Hi,
I create a ticket to formalize the development of this new feature. 
This ticket will be attached to the pull request.
The bookmarks will be completely redone to become observation lists.

Here the requirment :

 The Bookmarks tool will be deep refactored and visually splitted on 2 window - the first windows for using lists of bookmarked objects and the second list for adding/editing lists of objects. S
[Doc1.pdf](https://github.com/Stellarium/stellarium/files/4213417/Doc1.pdf)
[Doc1.pdf](https://github.com/Stellarium/stellarium/files/4213423/Doc1.pdf)
[Doc2.pdf](https://github.com/Stellarium/stellarium/files/4213424/Doc2.pdf)

ee sketches in the attachment.

Some comments:

- 1) We may store all lists in one file - for example observingLists.json - and this file just storing list of lists;  

- 2) Any list should storing 2 additional hidden properties for manage import/export of the data - unique ID and data of latest changes. Unique ID already implemented for current bookmarks.

- 3) Export and import tool should export/import list of bookmarks as currently implemented (subset of observingLists.json file)

- 4) All list shouldn't be stored translated data - english only names of designations. Translation should be applied on stage of visualization (creating table of view) data. Formal we may store only English name (or designation) and coordinates of object (currently in bookmarks we use coordinates to create a custom marker for object, who is not finded in Stellarium's objects) - and adding the coordinates (for finded objects - current coordinates), magnitude and abbreviation of constellation into table on stage of visualization of list.

- 5) "Empty" observingLists.json should store one list without objects - Bookmarks.

- 6) All files in UTF-8 and we may use special extension for separate lists (an exchange format of data) - .sol (stellarium observing list) or .skylist, or something else.

- 7) In future we may add tool for import observing lists from stellarium.org website (online import data) - of course if this tool will be very appropriated by users and we'll get feedback.

- 8) At the moment we store the location and date of observing data for each bookmark item and this is not very frequently use by users. But if we'll feedback we may add date and location data for all items of list as property for list: adding 2 fields into window 2 for define location and date/time and 2 checkboxes in first window for enable/disable applies those data.

- 9) Selecting object should be implemented by double click on list of object (+maybe adding separate button for go to object as currently implemented in Bookmarks tool) 

- 10) Maybe adding column "Type" (Type of object) will be useful. In any way we should use model to manage the data (like in Bookmarks at the moment).

I will update regulary this ticket to show development progress.
Currently the developpment is on going.

Jocelyn
[Doc1.pdf](https://github.com/Stellarium/stellarium/files/4213410/Doc1.pdf)
[Doc2.pdf](https://github.com/Stellarium/stellarium/files/4213412/Doc2.pdf)

### LINKED ISSUE #1884: B 287 (Barnard 287) is not bookmarkable
### Expected Behaviour
Barnard 287 appears in the bookmark list, a click on it will goto B 287 in Scorpius (close to M7).

### Actual Behaviour
Stellarium selects double star B 287 in Libra.
It is not possible to add Barnard 287 to the bookmarks.

### Steps to reproduce
Goto Barnard 287, add to bookmarks and click on it.

### System
* Stellarium version: 0.21.1
* Operating system: macos 11.4
* Graphics Card: -
* Screen type (if applicable): -

### Logfile 
If possible, attach the logfile `log.txt` from your user data directory. Look into the Guide for its location.

### LINKED ISSUE #1932: 0.21.2: observing lists core dump
test procedure:

1. open obs lists
2. create new list
3. type name
4. select a star
5. add object
6. do not close any dialog
6. immediately export list (indeed, in the object list creation mode)
7. then import the list

bug: core dump.

and this crash could have been prevented, it was predicted [here](https://github.com/Stellarium/stellarium/pull/1816#issuecomment-890382301). 

extensive discussion here: #976 

Operating System: Kubuntu 20.04
KDE Plasma Version: 5.18.5
KDE Frameworks Version: 5.68.0
Qt Version: 5.12.8
Kernel Version: 5.4.0-86-generic
OS Type: 64-bit
Processors: 8 × Intel® Core™ i7-10510U CPU @ 1.80GHz
Memory: 15.3 GiB of RAM

### LINKED ISSUE #1933: 0.21.2: obs list sends user to Null Island
1. add Sun to obs list (same issue with stars, probably with any object)
2. double click from within obs list to recenter the view on the object

bug: user location is reset.

and this bug could have been prevented, it was predicted [here](https://github.com/Stellarium/stellarium/pull/1816#issuecomment-890382301). 

extensive discussion here: #976 

Operating System: Kubuntu 20.04
KDE Plasma Version: 5.18.5
KDE Frameworks Version: 5.68.0
Qt Version: 5.12.8
Kernel Version: 5.4.0-86-generic
OS Type: 64-bit
Processors: 8 × Intel® Core™ i7-10510U CPU @ 1.80GHz
Memory: 15.3 GiB of RAM

### LINKED ISSUE #1941: Observing List: selecting list or use Edit button unselects the selected object
### Expected Behaviour
Observing List. When adding a new object to an new or existing list,  I press Edit and then can use the button Add Object to add the object and save.
Or selecting an object, open a existing list from drop-down and then Add to that list and save

### Actual Behaviour
When pressing the Edit button or opening an list from drop-down, the selected object gets unselected and Add Object does not work. I have to re-select the object with the Observing List Modification List window **open** and then the Add Object button works and can be saved.

### Steps to reproduce
- select/highlight an object
- open the Observing List
- select a list from drop down >> object is unselected, nothing to add.
   or
- make a new list and use Edit >> object is unselected, nothing to add.
- with Observing List Modification List window open, re-select the object and then Add is working

### System
* Stellarium version: 0.21.2
* Operating system: Windows 10 20H2
* Graphics Card: MSI RTX 2080 ti Gaming x Trio

### Logfile 
[log.txt](https://github.com/Stellarium/stellarium/files/7253755/log.txt)
.

### LINKED ISSUE #1944: Bookmarks can not be added/imported into Observing List
### Expected Behaviour
With upgrading from 0.21.1 to 0.21.2, Bookmarks should have converted into the Observing List or the import function of the Observing List should import Bookmarks (bookmarks.json). If I understand correctly, this is mentioned in #976 

### Actual Behaviour
After the upgrade the Observing List is empty and Bookmarks (from own system but also downloads from sites) can not be imported. Either nothing is happening or Stellarium is crashing (crash is Stellarium suddenly closing and log.txt has no data of the crash. See log below)

### Steps to reproduce
See above.

### System
* Stellarium version: 0.21.2
* Operating system: Windows 10 Pro 64bit 20H2 (19042.1237)
* Graphics Card: Graphics Card: MSI RTX 2080 ti Gaming x Trio

### Log file
[log.txt](https://github.com/Stellarium/stellarium/files/7263850/log.txt)
.

### LINKED ISSUE #2411: Segfault in Observing Lists
Moved from a different issue (#2232):

Like many others I am also interested in creating, exporting and importing observing lists in Stellarium and would like to see this feature become a reality. In trying to understand what's going on I just run Stellarium through gdb. I guess you developers do this all the time. At this point I cannot understand what is going wrong and cannot contribute a solution. But in case it is useful for you I post the backtrace below after a segfault caused by importing an observing list previously created in the same program. I will keep looking anyway.

OS: openSUSE Tumbleweed (rolling release), Stellarium version 0.22.1 (built from source)

```
Thread 1 "stellarium" received signal SIGSEGV, Segmentation fault.
0x00007fffedf86aa0 in QVariant::toString() const () from /lib64/libQt5Core.so.5
Missing separate debuginfos, use: zypper install Mesa-dri-nouveau-debuginfo-22.0.1-307.2.x86_64 Mesa-libGL1-debuginfo-22.0.1-307.1.x86_64 Mesa-libglapi0-debuginfo-22.0.1-307.1.x86_64 krb5-debuginfo-1.19.2-4.5.x86_64 libLLVM14-debuginfo-14.0.0-1.2.x86_64 libQt5Concurrent5-debuginfo-5.15.2+kde294-5.2.x86_64 libQt5Core5-debuginfo-5.15.2+kde294-5.2.x86_64 libQt5DBus5-debuginfo-5.15.2+kde294-5.2.x86_64 libQt5Gui5-debuginfo-5.15.2+kde294-5.2.x86_64 libQt5Positioning5-debuginfo-5.15.2+kde6-2.9.x86_64 libQt5PrintSupport5-debuginfo-5.15.2+kde294-5.2.x86_64 libQt5Widgets5-debuginfo-5.15.2+kde294-5.2.x86_64 libQtQuick5-debuginfo-5.15.2+kde43-2.1.x86_64 libX11-6-debuginfo-1.7.5-1.1.x86_64 libX11-xcb1-debuginfo-1.7.5-1.1.x86_64 libXcursor1-debuginfo-1.2.1-1.1.x86_64 libavcodec58_134-debuginfo-4.4.1-10.4.x86_64 libavformat58_76-debuginfo-4.4.1-10.4.x86_64 libavutil56_70-debuginfo-4.4.1-10.4.x86_64 libbluray2-debuginfo-1.3.1-1.1.x86_64 libdouble-conversion3-debuginfo-3.2.0-1.1.x86_64 libexpat1-debuginfo-2.4.8-1.1.x86_64 libfreetype6-debuginfo-2.12.0-1.1.x86_64 libgcc_s1-debuginfo-12.0.1+git192423-1.2.x86_64 libglib-2_0-0-debuginfo-2.72.0-2.1.x86_64 libgomp1-debuginfo-12.0.1+git192423-1.2.x86_64 libharfbuzz-subset0-debuginfo-4.2.0-1.2.x86_64 libharfbuzz0-debuginfo-4.2.0-1.2.x86_64 libicu70-debuginfo-71.1-1.1.x86_64 libjpeg8-debuginfo-8.2.2-68.1.x86_64 libminizip1-debuginfo-1.2.11-19.1.x86_64 libopenmpt0-debuginfo-0.6.2-1.1.x86_64 libqt5-qtwebengine-debuginfo-5.15.9-1.2.x86_64 libre2-9-debuginfo-20220401-45.1.x86_64 librtmp1-debuginfo-2.4.20151223.fa8646d-1.83.x86_64 libsndfile1-debuginfo-1.1.0-1.1.x86_64 libstdc++6-debuginfo-12.0.1+git192423-1.2.x86_64 libsystemd0-debuginfo-250.4-4.1.x86_64 libudev1-debuginfo-250.4-4.1.x86_64 libunistring2-debuginfo-1.0-1.1.x86_64 libva-drm2-debuginfo-2.14.0-1.2.x86_64 libva2-debuginfo-2.14.0-1.2.x86_64 libx264-161-debuginfo-0.161+git20200912.d198931a-2.30.x86_64 libx265-199-debuginfo-3.5-2.27.x86_64 libxvidcore4-debuginfo-1.3.7-1.36.x86_64 libz1-debuginfo-1.2.11-19.1.x86_64 mozilla-nss-debuginfo-3.76.1-2.1.x86_64
(gdb) backtrace 
#0  0x00007fffedf86aa0 in QVariant::toString() const () from /lib64/libQt5Core.so.5
#1  0x0000000000696be3 in ObsListCreateEditDialog::obsListImportListButtonPresssed() ()
#2  0x00007fffedf7d488 in ?? () from /lib64/libQt5Core.so.5
#3  0x00007fffef1950d2 in QAbstractButton::clicked(bool) () from /lib64/libQt5Widgets.so.5
#4  0x00007fffef19534a in ?? () from /lib64/libQt5Widgets.so.5
#5  0x00007fffef196f10 in ?? () from /lib64/libQt5Widgets.so.5
#6  0x00007fffef197133 in QAbstractButton::mouseReleaseEvent(QMouseEvent*) () from /lib64/libQt5Widgets.so.5
#7  0x00007fffef0e3cde in QWidget::event(QEvent*) () from /lib64/libQt5Widgets.so.5
#8  0x00007fffef0a1b0f in QApplicationPrivate::notify_helper(QObject*, QEvent*) () from /lib64/libQt5Widgets.so.5
#9  0x00007fffef0a9614 in QApplication::notify(QObject*, QEvent*) () from /lib64/libQt5Widgets.so.5
#10 0x00007fffedf46b5a in QCoreApplication::notifyInternal2(QObject*, QEvent*) () from /lib64/libQt5Core.so.5
#11 0x00007fffef0a8123 in QApplicationPrivate::sendMouseEvent(QWidget*, QMouseEvent*, QWidget*, QWidget*, QWidget**, QPointer<QWidget>&, bool, bool) () from /lib64/libQt5Widgets.so.5
#12 0x00007fffef3ca786 in ?? () from /lib64/libQt5Widgets.so.5
#13 0x00007fffef3b4608 in QGraphicsItem::sceneEvent(QEvent*) () from /lib64/libQt5Widgets.so.5
#14 0x00007fffef3d72e1 in ?? () from /lib64/libQt5Widgets.so.5
#15 0x00007fffef3dce5c in QGraphicsScene::mouseReleaseEvent(QGraphicsSceneMouseEvent*) ()
   from /lib64/libQt5Widgets.so.5
#16 0x00007fffef3e9a51 in QGraphicsScene::event(QEvent*) () from /lib64/libQt5Widgets.so.5
#17 0x00007fffef0a1b0f in QApplicationPrivate::notify_helper(QObject*, QEvent*) () from /lib64/libQt5Widgets.so.5
#18 0x00007fffedf46b5a in QCoreApplication::notifyInternal2(QObject*, QEvent*) () from /lib64/libQt5Core.so.5
#19 0x00007fffef407370 in QGraphicsView::mouseReleaseEvent(QMouseEvent*) () from /lib64/libQt5Widgets.so.5
#20 0x00007fffef0e3cde in QWidget::event(QEvent*) () from /lib64/libQt5Widgets.so.5
#21 0x00007fffef1905fe in QFrame::event(QEvent*) () from /lib64/libQt5Widgets.so.5
#22 0x00007fffedf468c2 in QCoreApplicationPrivate::sendThroughObjectEventFilters(QObject*, QEvent*) ()
   from /lib64/libQt5Core.so.5
#23 0x00007fffef0a1afe in QApplicationPrivate::notify_helper(QObject*, QEvent*) () from /lib64/libQt5Widgets.so.5
#24 0x00007fffef0a9614 in QApplication::notify(QObject*, QEvent*) () from /lib64/libQt5Widgets.so.5
#25 0x00007fffedf46b5a in QCoreApplication::notifyInternal2(QObject*, QEvent*) () from /lib64/libQt5Core.so.5
#26 0x00007fffef0a8123 in QApplicationPrivate::sendMouseEvent(QWidget*, QMouseEvent*, QWidget*, QWidget*, QWidget**, QPointer<QWidget>&, bool, bool) () from /lib64/libQt5Widgets.so.5
#27 0x00007fffef0fcddc in ?? () from /lib64/libQt5Widgets.so.5
#28 0x00007fffef1001f5 in ?? () from /lib64/libQt5Widgets.so.5
#29 0x00007fffef0a1b0f in QApplicationPrivate::notify_helper(QObject*, QEvent*) () from /lib64/libQt5Widgets.so.5
#30 0x00007fffedf46b5a in QCoreApplication::notifyInternal2(QObject*, QEvent*) () from /lib64/libQt5Core.so.5
--Type <RET> for more, q to quit, c to continue without paging--
#31 0x00007fffee848a77 in QGuiApplicationPrivate::processMouseEvent(QWindowSystemInterfacePrivate::MouseEvent*) ()
   from /lib64/libQt5Gui.so.5
#32 0x00007fffee81e86c in QWindowSystemInterface::sendWindowSystemEvents(QFlags<QEventLoop::ProcessEventsFlag>) ()
   from /lib64/libQt5Gui.so.5
#33 0x00007fffe45a1e9a in ?? () from /lib64/libQt5XcbQpa.so.5
#34 0x00007fffea449122 in g_main_context_dispatch () from /lib64/libglib-2.0.so.0
#35 0x00007fffea4494b8 in ?? () from /lib64/libglib-2.0.so.0
#36 0x00007fffea44956f in g_main_context_iteration () from /lib64/libglib-2.0.so.0
#37 0x00007fffedf9e2b4 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) ()
   from /lib64/libQt5Core.so.5
#38 0x00007fffedf4555b in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () from /lib64/libQt5Core.so.5
#39 0x00007fffedf4d820 in QCoreApplication::exec() () from /lib64/libQt5Core.so.5
#40 0x0000000000504def in main ()
```

_Originally posted by @dhdeangelis in https://github.com/Stellarium/stellarium/issues/2232#issuecomment-1101384793_

### LINKED ISSUE #2525: Add additional names to observing list
**Is your feature request related to a problem? Please describe.**

I can't always remember what an object is named from its catalog number, or what its number is based on the name. Unfortunately I couldn't find a way to add additional names to the observing list.

**Describe the solution you'd like**

Add an additional column for names if the user wants to see them.

**Describe alternatives you've considered**

Tried moving the column dividers, right clicking all around the observing list to see if this was already enabled.

**Additional context**

Additional names for the Pelican Nebula
Screenshot_2022-07-20_10-27-13

Only one catalog name visible in the observing list.
Screenshot_2022-07-20_10-14-20

### PR DESCRIPTION
**Pull request by Jocelyn1109**
Fixes #2525
Fixes #2411
Fixes #1944
Fixes #1941
Fixes #1933
Fixes #1932
Fixes #1884
Fixes #976

### CODE REVIEW

**Comment by alex-w** [comment]
@Jocelyn1109 please use QRegularExpression

**Review by alex-w — CHANGES_REQUESTED** [review]
In the editor of list I see 2 checkboxes - Ra & Dec - why not use just one - Coordinates?

The Name column - I think objects without names should contain "dash" char.

The Magnitude column: please use formatted output (2 digits after dot for all magnitudes + "dash" char for magnitudes over 98 - because magnitude 99 used for objects without magnitudes).

P.S. See AstroCalc tools for obtain examples for use "dash" char and formatting output of magnitudes

**Comment by Jocelyn1109** [comment]
@alex-w  "In the editor of list I see 2 checkboxes - Ra & Dec - why not use just one - Coordinates?" -> yes, we can use ony one checkbox for both if you want. I thought they should be separated.

**Comment by alex-w** [comment]
@Jocelyn1109 please mark which issues this pull request should closes (via edit description of pull request)

**Comment by alex-w** [comment]
I see no fixes for Name and Magnitude columns...

**Comment by Jocelyn1109** [comment]
I forgot I will aff the fix this week-end. Sorry

Envoyé depuis Yahoo Mail pour Android 
 
  Le ven., sept. 9, 2022 à 22:12, Alexander V. ***@***.***> a écrit:   


I see no fixes for Name and Magnitude columns...

**Comment by Jocelyn1109** [comment]
@alex-w  for the empty name, do you speak about "nameI18n" ?

**Comment by alex-w** [comment]
> @alex-w for the empty name, do you speak about "nameI18n" ?

Yes, see selected line on the screenshot

stellarium-018

**Comment by Jocelyn1109** [comment]
> 

and for magnitude it is like that:
`magnitude > 98.f ? dash : QString::number(magnitude, 'f', 2)));`

**Review by alex-w — APPROVED** [review]

**Comment by alex-w** [comment]
I think the changes in the code is ready to public testing

**Comment by gzotti** [comment]
I have a few entries for the Moon, stored with date/location. In the bookmark list, all show the coordinates of the "current" moon. Would it not be better to store&retrieve also coordinates, or at least recreate the coordinates from the stored time&locations?

When a bookmark was stored, it apparently stored the location name, but no other data. Sometimes this location comes from a "landscape". On bookmark loading, users are thrown to long/lat 0/0. If location was not found, it is better to not move to the ocean around the equator. 

On the plus side, I have so far not triggered a crash :-)

**Comment by Jocelyn1109** [comment]
Hi, @gzotti is it possible to have your observinglist.json ?
I would like to analyze and understand what you want .  :)

Many thanks.

Jocelyn

**Comment by gzotti** [comment]
OK, here. (Renamed for Github)

[observingList.json.txt](https://github.com/Stellarium/stellarium/files/9549304/observingList.json.txt)

I see when location is empty, the user is moved to 0/0. It would be better to not move location in this case. Probably also locations only retrieved from the landscapes are lost on storing when there is no coordinates for the Landscape placename stored in the location database. It may be useful to explore storing the whole location (name, coordinates, timezone, LP info, ...) into a string which could be retrieved by Location::createFromLine(). For this, developing a fitting Location::toString() would be required. In any case, if no numerical data can be found for a location, the position should not change to just 0/0. (Note somebody may store 0/0 explicitly! Then it is required to go there...)

**Comment by gzotti** [comment]
Of course StelLocation::serializeToLine() is the inverse of createFromLine() which you can use to encode all location data into a single entry for the JSON.

**Comment by Jocelyn1109** [comment]
@gzotti 
"I have a few entries for the Moon, stored with date/location. In the bookmark list, all show the coordinates of the "current" moon. Would it not be better to store&retrieve also coordinates, or at least recreate the coordinates from the stored time&locations?" -> yes it is absolutely better and I will modify this point :)

**Comment by gzotti** [comment]
Yes, this is good. Imagine you want to show the eclipsed Moon from various locations at the same time. This should allow demonstrating that it appears equally dark, but in different places (az/alt) in the sky, and esp. in slightly differing places depending on topographic parallax.

As said, please also consider optionally storing the landscape ID for an automated context change.

**Comment by Jocelyn1109** [comment]
@gzotti 
the landscape id is already taken into account but optional (checkbox in create dialog).

```
"observingLists": {
        "{4c48abf7-8ea3-4b06-be90-dc7655b5e2d0}": {
            "creation date": "2022-08-08 07:00:45",
            "description": "location tests",
            "landscape id": "",
            "name": "location test",
            "objects": [
                {
                    "constellation": "CMa",
                    "dec": "-26°23'26\"",
                    "designation": "HIP 34444",
                    "fov": 60,
                    "isVisibleMarker": false,
                    "jd": 2459799.948451157,
                    "location": "Tamiyah, Northern Africa",
                    "magnitude": "1.8",
                    "nameI18n": "Wezen",
                    "objtype": "double star",
                    "ra": "7h08m22.2s",
                    "type": "Star"
                },
            ],
            "sorting": ""
        },
```

**Comment by alex-w** [comment]
@Jocelyn1109 is it possible pushing the fixes in next 2 days to adding them into RC1?

**Comment by Jocelyn1109** [comment]
@alex-w yes I will do my best :)

**Comment by alex-w** [comment]
@gzotti is the currently state of the pull request enough to merge and extract the translatable lines to prepare RC1 tomorrow?

**Comment by Jocelyn1109** [comment]
@alex-w @gzotti 
Tonight I will push the code that fixes a SIGSEV.
But for the moment the problem of Georg is not corrected because I do not yet know how to retrieve the coordinates (Ra and Dec) of an object with the time and the location.

**Comment by gzotti** [comment]
> @alex-w @gzotti Tonight I will push the code that fixes a SIGSEV. But for the moment the problem of Georg is not corrected because I do not yet know how to retrieve the coordinates (Ra and Dec) of an object with the time and the location.

If this is a problem only for the old bookmark lists, don't bother.  For new entries it should be enough to store Ra/Dec when the entry is created with timestamp. But it is important to not change location if the location cannot be decoded correctly. 

Hmm, location should as discussed be stored as StelLocation::serializeToLine() and decoded with StelLocation::createFromLine(const QString& line); to fully store data in case location comes from a landscape or was manually configured.

The GUI should be edited a bit: move all optionally stored items above or below. Currently checkboxes in the list edit panel to store coordinates and fov are below, landscape is above, and there is no JD entry.  In the data retrieval panel, there are only checkboxes for JD and location, no landscapes and no fov. 

Please also rebase this branch from current master.

File: stellarium.kdev4 [hunk 1]
```diff
@@ -0,0 +1,4 @@
+[Project]
```
Comment (gzotti) [inline stellarium.kdev4:1]: I think this file should be removed from the repository.

**Pull request merged**